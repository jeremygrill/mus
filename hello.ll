; ModuleID = 'Mus'

%chord_node = type <{ i32, i32* }>

@fmt = private unnamed_addr constant [4 x i8] c"%d\0A\00"
@fmt.1 = private unnamed_addr constant [12 x i8] c"(%d,%d|%d)\0A\00"
@fmt.2 = private unnamed_addr constant [14 x i8] c"[(%d,%d|%d)]\0A\00"

declare i32 @printf(i8*, ...)

declare i32 @play(i8*, ...)

define i32 @main() {
entry:
  %i = alloca i32
  %j = alloca i32
  %b = alloca i1
  %a = alloca i32
  %n = alloca i32
  %c = alloca %chord_node*
  %d = alloca %chord_node*
  store i32 1, i32* %i
  store i32 1694629891, i32* %a
  store i32 1778843656, i32* %n
  %c1 = alloca %chord_node*
  %malloccall = tail call i8* @malloc(i32 ptrtoint (%chord_node* getelementptr (%chord_node, %chord_node* null, i32 1) to i32))
  %a2 = bitcast i8* %malloccall to %chord_node*
  store %chord_node* %a2, %chord_node** %c1
  %a3 = load %chord_node*, %chord_node** %c1
  %a4 = getelementptr inbounds %chord_node, %chord_node* %a3, i32 0, i32 0
  %n1 = load i32, i32* %n
  store i32 %n1, i32* %a4
  %a5 = load %chord_node*, %chord_node** %c1
  %a6 = getelementptr inbounds %chord_node, %chord_node* %a5, i32 0, i32 1
  store i32* null, i32** %a6
  %a7 = load %chord_node*, %chord_node** %c1
  %malloccall.2 = tail call i8* @malloc(i32 ptrtoint (%chord_node* getelementptr (%chord_node, %chord_node* null, i32 1) to i32))
  %a23 = bitcast i8* %malloccall.2 to %chord_node*
  store %chord_node* %a23, %chord_node** %c1
  %a34 = load %chord_node*, %chord_node** %c1
  %a45 = getelementptr inbounds %chord_node, %chord_node* %a34, i32 0, i32 0
  %n6 = load i32, i32* %n
  store i32 %n6, i32* %a45
  %a57 = load %chord_node*, %chord_node** %c1
  %a68 = getelementptr inbounds %chord_node, %chord_node* %a57, i32 0, i32 1
  store i32* %a4, i32** %a68
  %a79 = load %chord_node*, %chord_node** %c1
  %final = load %chord_node*, %chord_node** %c1
  store %chord_node* %final, %chord_node** %c
  %n10 = load i32, i32* %n
  %tmp = sdiv i32 %n10, 16777216
  %tmp11 = and i32 %n10, 16777215
  %tmp12 = sdiv i32 %tmp11, 65536
  %tmp13 = and i32 %n10, 65535
  %printn = call i32 (i8*, ...) @printf(i8* getelementptr inbounds ([12 x i8], [12 x i8]* @fmt.1, i32 0, i32 0), i32 %tmp, i32 %tmp12, i32 %tmp13)
  %a14 = load i32, i32* %a
  %tmp15 = sdiv i32 %a14, 16777216
  %tmp16 = and i32 %a14, 16777215
  %tmp17 = sdiv i32 %tmp16, 65536
  %tmp18 = and i32 %a14, 65535
  %printn19 = call i32 (i8*, ...) @printf(i8* getelementptr inbounds ([12 x i8], [12 x i8]* @fmt.1, i32 0, i32 0), i32 %tmp15, i32 %tmp17, i32 %tmp18)
  %b1 = alloca i32
  %tmp20 = alloca %chord_node*
  store i32 0, i32* %b1
  %c21 = load %chord_node*, %chord_node** %c
  store %chord_node* %c21, %chord_node** %tmp20
  %b3 = load %chord_node*, %chord_node** %tmp20
  %b4 = getelementptr inbounds %chord_node, %chord_node* %b3, i32 0, i32 0
  %b5 = load i32, i32* %b4
  %tmp22 = sdiv i32 %b5, 16777216
  %tmp23 = and i32 %b5, 16777215
  %tmp24 = sdiv i32 %tmp23, 65536
  %tmp25 = and i32 %b5, 65535
  %printf = call i32 (i8*, ...) @printf(i8* getelementptr inbounds ([14 x i8], [14 x i8]* @fmt.2, i32 0, i32 0), i32 %tmp22, i32 %tmp24, i32 %tmp25)
  %b7 = load %chord_node*, %chord_node** %tmp20
  ret i32 0
}

declare noalias i8* @malloc(i32)
