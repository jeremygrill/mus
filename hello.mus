Int run()
{
	Int i;
	Int j;
	Bool b;
	Note a;
	Note n;
	Chord c;
	Chord d;
	4 + 2;
	i = 1; 
	b = false;
	a = (1,2|3);
	n = (6,7|8);
	d[a a a a a];
	c[a n];
	print(88);
	print(i);
	c;
	d;
	
	while(b){
		print(33);	
	}

	return 0;

}
