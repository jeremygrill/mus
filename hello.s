	.section	__TEXT,__text,regular,pure_instructions
	.macosx_version_min 10, 10
	.globl	_main                   ## -- Begin function main
	.p2align	4, 0x90
_main:                                  ## @main
	.cfi_startproc
## BB#0:                                ## %entry
	pushq	%rbx
Lcfi0:
	.cfi_def_cfa_offset 16
	subq	$64, %rsp
Lcfi1:
	.cfi_def_cfa_offset 80
Lcfi2:
	.cfi_offset %rbx, -16
	movl	$1, 36(%rsp)
	movl	$1694629891, 20(%rsp)   ## imm = 0x65020003
	movl	$1778843656, 4(%rsp)    ## imm = 0x6A070008
	movl	$12, %edi
	callq	_malloc
	movq	%rax, %rbx
	movq	%rbx, 8(%rsp)
	movl	4(%rsp), %eax
	movl	%eax, (%rbx)
	movq	8(%rsp), %rax
	movq	$0, 4(%rax)
	movl	$12, %edi
	callq	_malloc
	movq	%rax, 8(%rsp)
	movl	4(%rsp), %ecx
	movl	%ecx, (%rax)
	movq	8(%rsp), %rax
	movq	%rbx, 4(%rax)
	movq	8(%rsp), %rax
	movq	%rax, 24(%rsp)
	movl	4(%rsp), %eax
	movl	%eax, %esi
	sarl	$31, %esi
	shrl	$8, %esi
	addl	%eax, %esi
	sarl	$24, %esi
	movzwl	%ax, %ecx
	movl	%eax, %edx
	andl	$16711680, %edx         ## imm = 0xFF0000
	shrl	$16, %edx
	leaq	L_fmt.1(%rip), %rbx
	xorl	%eax, %eax
	movq	%rbx, %rdi
	callq	_printf
	movl	20(%rsp), %eax
	movl	%eax, %esi
	sarl	$31, %esi
	shrl	$8, %esi
	addl	%eax, %esi
	sarl	$24, %esi
	movzwl	%ax, %ecx
	movl	%eax, %edx
	andl	$16711680, %edx         ## imm = 0xFF0000
	shrl	$16, %edx
	xorl	%eax, %eax
	movq	%rbx, %rdi
	callq	_printf
	movl	$0, 32(%rsp)
	movq	24(%rsp), %rax
	movq	%rax, 40(%rsp)
	movl	(%rax), %eax
	movl	%eax, %esi
	sarl	$31, %esi
	shrl	$8, %esi
	addl	%eax, %esi
	sarl	$24, %esi
	movzwl	%ax, %ecx
	movl	%eax, %edx
	andl	$16711680, %edx         ## imm = 0xFF0000
	shrl	$16, %edx
	leaq	L_fmt.2(%rip), %rdi
	xorl	%eax, %eax
	callq	_printf
	xorl	%eax, %eax
	addq	$64, %rsp
	popq	%rbx
	retq
	.cfi_endproc
                                        ## -- End function
	.section	__TEXT,__cstring,cstring_literals
L_fmt:                                  ## @fmt
	.asciz	"%d\n"

L_fmt.1:                                ## @fmt.1
	.asciz	"(%d,%d|%d)\n"

L_fmt.2:                                ## @fmt.2
	.asciz	"[(%d,%d|%d)]\n"


.subsections_via_symbols
