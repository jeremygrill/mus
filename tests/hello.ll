; ModuleID = 'Mus'
source_filename = "Mus"

%chord_node = type { i32, %chord_node* }

@fmt = private unnamed_addr constant [4 x i8] c"%d\0A\00"
@fmt.1 = private unnamed_addr constant [12 x i8] c"(%d,%d|%d)\0A\00"

declare i32 @printf(i8*, ...)

declare i32 @printc(%chord_node*)

declare i32 @play(i32)

define i32 @main() {
entry:
  %a = alloca i32
  %n = alloca i32
  %suh = alloca i32
  %c = alloca %chord_node*
  %d = alloca %chord_node*
  store i32 184614923, i32* %a
  store i32 50528259, i32* %suh
  store i32 34996226, i32* %n
  %c1 = alloca %chord_node*
  %malloccall = tail call i8* @malloc(i32 ptrtoint (%chord_node* getelementptr (%chord_node, %chord_node* null, i32 1) to i32))
  %a2 = bitcast i8* %malloccall to %chord_node*
  store %chord_node* %a2, %chord_node** %c1
  %a3 = load %chord_node*, %chord_node** %c1
  %a4 = getelementptr inbounds %chord_node, %chord_node* %a3, i32 0, i32 0
  %suh1 = load i32, i32* %suh
  store i32 %suh1, i32* %a4
  %a5 = load %chord_node*, %chord_node** %c1
  %a6 = getelementptr inbounds %chord_node, %chord_node* %a5, i32 0, i32 1
  store %chord_node* null, %chord_node** %a6
  %a7 = load %chord_node*, %chord_node** %c1
  %c12 = alloca %chord_node*
  %malloccall3 = tail call i8* @malloc(i32 ptrtoint (%chord_node* getelementptr (%chord_node, %chord_node* null, i32 1) to i32))
  %a24 = bitcast i8* %malloccall3 to %chord_node*
  store %chord_node* %a24, %chord_node** %c12
  %a35 = load %chord_node*, %chord_node** %c12
  %a46 = getelementptr inbounds %chord_node, %chord_node* %a35, i32 0, i32 0
  %n7 = load i32, i32* %n
  store i32 %n7, i32* %a46
  %a58 = load %chord_node*, %chord_node** %c12
  %a69 = getelementptr inbounds %chord_node, %chord_node* %a58, i32 0, i32 1
  store %chord_node* %a7, %chord_node** %a69
  %a710 = load %chord_node*, %chord_node** %c12
  %c111 = alloca %chord_node*
  %malloccall12 = tail call i8* @malloc(i32 ptrtoint (%chord_node* getelementptr (%chord_node, %chord_node* null, i32 1) to i32))
  %a213 = bitcast i8* %malloccall12 to %chord_node*
  store %chord_node* %a213, %chord_node** %c111
  %a314 = load %chord_node*, %chord_node** %c111
  %a415 = getelementptr inbounds %chord_node, %chord_node* %a314, i32 0, i32 0
  %a16 = load i32, i32* %a
  store i32 %a16, i32* %a415
  %a517 = load %chord_node*, %chord_node** %c111
  %a618 = getelementptr inbounds %chord_node, %chord_node* %a517, i32 0, i32 1
  store %chord_node* %a710, %chord_node** %a618
  %a719 = load %chord_node*, %chord_node** %c111
  store %chord_node* %a719, %chord_node** %c
  %a20 = load i32, i32* %a
  %n21 = load i32, i32* %n
  %c122 = alloca %chord_node*
  %malloccall23 = tail call i8* @malloc(i32 ptrtoint (%chord_node* getelementptr (%chord_node, %chord_node* null, i32 1) to i32))
  %a224 = bitcast i8* %malloccall23 to %chord_node*
  store %chord_node* %a224, %chord_node** %c122
  %a325 = load %chord_node*, %chord_node** %c122
  %a426 = getelementptr inbounds %chord_node, %chord_node* %a325, i32 0, i32 0
  %n27 = load i32, i32* %n
  store i32 %n27, i32* %a426
  %a528 = load %chord_node*, %chord_node** %c122
  %a629 = getelementptr inbounds %chord_node, %chord_node* %a528, i32 0, i32 1
  store %chord_node* null, %chord_node** %a629
  %a730 = load %chord_node*, %chord_node** %c122
  %c131 = alloca %chord_node*
  %malloccall32 = tail call i8* @malloc(i32 ptrtoint (%chord_node* getelementptr (%chord_node, %chord_node* null, i32 1) to i32))
  %a233 = bitcast i8* %malloccall32 to %chord_node*
  store %chord_node* %a233, %chord_node** %c131
  %a334 = load %chord_node*, %chord_node** %c131
  %a435 = getelementptr inbounds %chord_node, %chord_node* %a334, i32 0, i32 0
  %a36 = load i32, i32* %a
  store i32 %a36, i32* %a435
  %a537 = load %chord_node*, %chord_node** %c131
  %a638 = getelementptr inbounds %chord_node, %chord_node* %a537, i32 0, i32 1
  store %chord_node* %a730, %chord_node** %a638
  %a739 = load %chord_node*, %chord_node** %c131
  store %chord_node* %a739, %chord_node** %d
  %c40 = load %chord_node*, %chord_node** %c
  %printc = call i32 @printc(%chord_node* %c40)
  %d41 = load %chord_node*, %chord_node** %d
  %printc42 = call i32 @printc(%chord_node* %d41)
  %a43 = load i32, i32* %a
  %play = call i32 @play(i32 %a43)
  ret i32 0
}

declare noalias i8* @malloc(i32)
