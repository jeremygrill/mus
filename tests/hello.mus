Int main()
{
	Int i;
	Int j;
	Bool b;
	Note a;
	Note n;
	Chord c;
	Chord d;
	4 + 2;
	i = 1; 
	//b = true;
	a = (11,1|11);
	n = (2,22|2);
	//d = [a a a a a];
	c = [a n a];
	//print(88);
	printn(n);
	printn(a);
	printc(c);
	//play(a);
	//c;
	//d;
	return 0;
}
