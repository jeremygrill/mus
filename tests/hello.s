	.section	__TEXT,__text,regular,pure_instructions
	.macosx_version_min 10, 11
	.globl	_main                   ## -- Begin function main
	.p2align	4, 0x90
_main:                                  ## @main
	.cfi_startproc
## %bb.0:                               ## %entry
	subq	$40, %rsp
	.cfi_def_cfa_offset 48
	movl	$1, 4(%rsp)
	movl	$16908291, 12(%rsp)     ## imm = 0x1020003
	movl	$101122056, 8(%rsp)     ## imm = 0x6070008
	movl	$40, %edi
	callq	_malloc
	movl	$16, %edi
	callq	_malloc
	movl	4(%rsp), %esi
	leaq	L_fmt(%rip), %rdi
	xorl	%eax, %eax
	callq	_printf
	xorl	%eax, %eax
	addq	$40, %rsp
	retq
	.cfi_endproc
                                        ## -- End function
	.section	__TEXT,__cstring,cstring_literals
L_fmt:                                  ## @fmt
	.asciz	"%d\n"

L_fmt.1:                                ## @fmt.1
	.asciz	"(%d,%d|%d)\n"

L_fmt.2:                                ## @fmt.2
	.asciz	"%g\n"


.subsections_via_symbols
