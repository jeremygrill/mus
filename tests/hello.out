(106,7|8)
(101,2|3)
[(101,2|3)(106,7|8)]
